Client/mods/deathmatch/logic/CClientModelManager.cpp

```diff
@@ -10,45 +10,38 @@
 
 #include "StdInc.h"
 
-CClientModelManager::CClientModelManager(CClientManager* pManager)
-{
-    for (ushort i = 0; i < MAX_MODEL_ID; i++)
-    {
-        m_Models[i] = nullptr;
-    }
-}
-
 CClientModelManager::~CClientModelManager(void)
 {
-    // Delete all our models
     RemoveAll();
 }
 
 void CClientModelManager::RemoveAll(void)
 {
     for (int i = 0; i < MAX_MODEL_ID; i++)
     {
-        Remove(m_Models[i]);
+        m_Models[i] = nullptr;
     }
     m_modelCount = 0;
 }
 
-void CClientModelManager::Add(CClientModel* pModel)
+void CClientModelManager::Add(const std::shared_ptr<CClientModel>& pModel)
 {
     if (m_Models[pModel->GetModelID()] != nullptr)
     {
-        dassert(m_Models[pModel->GetModelID()] == pModel);
+        dassert(m_Models[pModel->GetModelID()].get() == pModel.get());
         return;
     }
     m_Models[pModel->GetModelID()] = pModel;
     m_modelCount++;
 }
 
-bool CClientModelManager::Remove(CClientModel* pModel)
+bool CClientModelManager::Remove(const std::shared_ptr<CClientModel>& pModel)
 {
-    if (pModel && m_Models[pModel->GetModelID()] != nullptr)
+    int modelId = pModel ? pModel->GetModelID() : 0;
+    if (pModel && m_Models[modelId] != nullptr)
     {
-        m_Models[pModel->GetModelID()] = nullptr;
+        m_Models[modelId]->RestoreEntitiesUsingThisModel();
+        m_Models[modelId] = nullptr;
```

Comment:
Something looks weird here. Should this just return false if pModel is null?

Reply:
Ah yes, indeed

Reply:
Good catch. This is from the original code. It doesn't cause any serious bugs, although everything will work as intended, however, it can affect the return value of `engineFreeModel`.